# Kings County Real Estate - Housing Analysis

## Introduction:

Microsoft is thin

## Business Question:

We are trying to predict at what price your home would be valued at? 

## Data Questions:

1. Movies 



## Data - Importing & Cleaning:

###### Ratings:Genre Data - title.basics.csv and title.ratings.csv - Cleaning & Creating Joined Dataframe

This created dataframe that will show the dataset utiltized, downloaded from the link below.

https://osf.io/twq9p/

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

from bs4 import BeautifulSoup
import requests
import re

##### The following is a function to download via pandas csv, excel, or json files to jupyter notebook:

In [2]:
def files_import_pd(file_path):
    if file_path.endswith("csv"):
        return pd.read_csv(file_path)
    if file_path.endswith("tsv"):
        return pd.read_csv(file_path, sep="\t")
    if file_path.endswith("xlsx"):
        return pd.read_excel(file_path)
    if file_path.endswith("json"):
        return pd.read_json(file_path)
    else:
        print("NOT CSV/TSV/EXCEL/JSON FILE")

In [3]:
df1 = files_import_pd(r"C:\Users\bigbenx3\2021_flatiron\flatiron_projects\housing_analysis_project\kc_house_data.csv")
df1.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

Good. Imported the first dataframe. Let's look at its contents.

In [4]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

So it appears there are no missing/ empty values.

However, it seems there are datatypes that need to be altered for example: date is an object. Price is also a float which will be hard to input.

## Data - Manipulation

In [5]:
df1["price"] = df1["price"].astype(int)

In [6]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  int32  
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

The "price" column values are all integers now.

In [7]:
df1.shape

(21613, 21)

21613 rows and  21 columns.

Now, we want to drop certain columns to simplify the model we will make. And before we do that, we will create a copy of this original data frame. 

In [8]:
df_drop_cols = df1.copy()

In [9]:
df_drop_cols.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  int32  
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

Alright, same dataframe still but just a copy. Now, to drop the columns below:

date, view, sqft_above, sqft_basement, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15

In [10]:
df_drop_cols = df1.drop(["date","view","sqft_above","sqft_basement","yr_renovated","lat","long","sqft_living15","sqft_lot15"],axis="columns")

In [11]:
df_drop_cols.head()

id   price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900         3       1.00         1180      5650     1.0   
1  6414100192  538000         3       2.25         2570      7242     2.0   
2  5631500400  180000         2       1.00          770     10000     1.0   
3  2487200875  604000         4       3.00         1960      5000     1.0   
4  1954400510  510000         3       2.00         1680      8080     1.0   

   waterfront  condition  grade  yr_built  zipcode  
0           0          3      7      1955    98178  
1           0          3      7      1951    98125  
2           0          3      6      1933    98028  
3           0          5      7      1965    98136  
4           0          3      8      1987    98074

In [12]:
df_drop_cols.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           21613 non-null  int64  
 1   price        21613 non-null  int32  
 2   bedrooms     21613 non-null  int64  
 3   bathrooms    21613 non-null  float64
 4   sqft_living  21613 non-null  int64  
 5   sqft_lot     21613 non-null  int64  
 6   floors       21613 non-null  float64
 7   waterfront   21613 non-null  int64  
 8   condition    21613 non-null  int64  
 9   grade        21613 non-null  int64  
 10  yr_built     21613 non-null  int64  
 11  zipcode      21613 non-null  int64  
dtypes: float64(2), int32(1), int64(9)
memory usage: 1.9 MB


## Data Manipulation plus Cleaning

In [13]:
df_drop_cols.isnull().sum()

id             0
price          0
bedrooms       0
bathrooms      0
sqft_living    0
sqft_lot       0
floors         0
waterfront     0
condition      0
grade          0
yr_built       0
zipcode        0
dtype: int64

No null values found.

Now we want to see what kind of values we find in different columns that seems to be a lot of similar values. We want to see the unique values to get a better sense of the "spectrum", if you will.

In [14]:
import pandas as pd
df_drop_cols_unique_bedrooms = df_drop_cols["bedrooms"].unique()
print(sorted(df_drop_cols_unique_bedrooms))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 33]


So the "33 bedrooms" sticks out and is potential for error, so we will just match that to the corresponding square footage of the living space "sqft_living".

In [15]:
df_drop_cols[df_drop_cols.bedrooms>9]

id    price  bedrooms  bathrooms  sqft_living  sqft_lot  \
8757   1773100755   520000        11       3.00         3000      4960   
13314   627300145  1148000        10       5.25         4590     10920   
15161  5566100170   650000        10       2.00         3610     11914   
15870  2402100895   640000        33       1.75         1620      6000   
19254  8812401450   660000        10       3.00         2920      3745   

       floors  waterfront  condition  grade  yr_built  zipcode  
8757      2.0           0          3      7      1918    98106  
13314     1.0           0          3      9      2008    98004  
15161     2.0           0          4      7      1958    98006  
15870     1.0           0          5      7      1947    98103  
19254     2.0           0          4      7      1913    98105

Again, we see there to be a bit of a discrepancy with the "33 bedrooms" taking 1620 square ft of living space. 
This indicates that we need to do further cleaning in eliminating outliers. 

#### OUTLIER ALERT!!! (cell above)

In [16]:
import pandas as pd
df_drop_cols_unique_bathrooms = df_drop_cols["bathrooms"].unique()
print(sorted(df_drop_cols_unique_bathrooms))

[0.0, 0.5, 0.75, 1.0, 1.25, 1.5, 1.75, 2.0, 2.25, 2.5, 2.75, 3.0, 3.25, 3.5, 3.75, 4.0, 4.25, 4.5, 4.75, 5.0, 5.25, 5.5, 5.75, 6.0, 6.25, 6.5, 6.75, 7.5, 7.75, 8.0]


It turns out there are specific features that are found in .75 and .25 bathrooms. https://www.badeloftusa.com/buying-guides/75-bathroom/

In [17]:
import pandas as pd
df_drop_cols_unique_floors = df_drop_cols["floors"].unique()
print(sorted(df_drop_cols_unique_floors))

[1.0, 1.5, 2.0, 2.5, 3.0, 3.5]


It turns out it's possible to have half a floor, referring to https://rct9.msbexpress.net/RCT44Help/1/Content/Style_2_5_Story.htm#:~:text=A%202.5%20story%20is%20a,of%20the%20ground%20floor%20area.

In [18]:
import pandas as pd
df_drop_cols_unique_waterfront = df_drop_cols["waterfront"].unique()
print(sorted(df_drop_cols_unique_waterfront))

[0, 1]


"0" seems the absent of and "1" seems to be presence of a water front.

In [19]:
import pandas as pd
df_drop_cols_unique_condition = df_drop_cols["condition"].unique()
print(sorted(df_drop_cols_unique_condition))

[1, 2, 3, 4, 5]


In [20]:
import pandas as pd
df_drop_cols_unique_grade = df_drop_cols["grade"].unique()
print(sorted(df_drop_cols_unique_grade))

[1, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]


In [21]:
import pandas as pd
df_drop_cols_unique_yr_built = df_drop_cols["yr_built"].unique()
print(sorted(df_drop_cols_unique_yr_built))

[1900, 1901, 1902, 1903, 1904, 1905, 1906, 1907, 1908, 1909, 1910, 1911, 1912, 1913, 1914, 1915, 1916, 1917, 1918, 1919, 1920, 1921, 1922, 1923, 1924, 1925, 1926, 1927, 1928, 1929, 1930, 1931, 1932, 1933, 1934, 1935, 1936, 1937, 1938, 1939, 1940, 1941, 1942, 1943, 1944, 1945, 1946, 1947, 1948, 1949, 1950, 1951, 1952, 1953, 1954, 1955, 1956, 1957, 1958, 1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967, 1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015]


In [22]:
import pandas as pd
df_drop_cols_unique_zipcode = df_drop_cols["zipcode"].unique()
print(sorted(df_drop_cols_unique_zipcode))

[98001, 98002, 98003, 98004, 98005, 98006, 98007, 98008, 98010, 98011, 98014, 98019, 98022, 98023, 98024, 98027, 98028, 98029, 98030, 98031, 98032, 98033, 98034, 98038, 98039, 98040, 98042, 98045, 98052, 98053, 98055, 98056, 98058, 98059, 98065, 98070, 98072, 98074, 98075, 98077, 98092, 98102, 98103, 98105, 98106, 98107, 98108, 98109, 98112, 98115, 98116, 98117, 98118, 98119, 98122, 98125, 98126, 98133, 98136, 98144, 98146, 98148, 98155, 98166, 98168, 98177, 98178, 98188, 98198, 98199]


In [23]:
len(sorted(df_drop_cols_unique_zipcode))

70

So for almost all the columns, we have an idea based on the unique values.

Now, we also want to add a column called "price_per_sqft" going based off of the price of the home divided by square footage of the home ("sqft_living" = sqft_livingsquare - footage of the home)  https://www.homelight.com/blog/buyer-how-to-calculate-price-per-square-foot/

## QUESTION!!!

Is the price of a house including the square footage of the surrounding areas?
https://www.reinbrechthomes.com/blog/dont-get-hung-up-on-the-cost-per-square-foot-to-build-a-house/
https://www.quickenloans.com/learn/price-per-square-foot

I guess it's just the sqft_living space that should be concerned.

In [24]:
df3 = df_drop_cols.copy()
df3["price_per_sqft"] = df3["price"]/df3["sqft_living"]
df3.head()

id   price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900         3       1.00         1180      5650     1.0   
1  6414100192  538000         3       2.25         2570      7242     2.0   
2  5631500400  180000         2       1.00          770     10000     1.0   
3  2487200875  604000         4       3.00         1960      5000     1.0   
4  1954400510  510000         3       2.00         1680      8080     1.0   

   waterfront  condition  grade  yr_built  zipcode  price_per_sqft  
0           0          3      7      1955    98178      188.050847  
1           0          3      7      1951    98125      209.338521  
2           0          3      6      1933    98028      233.766234  
3           0          5      7      1965    98136      308.163265  
4           0          3      8      1987    98074      303.571429

In [25]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              21613 non-null  int64  
 1   price           21613 non-null  int32  
 2   bedrooms        21613 non-null  int64  
 3   bathrooms       21613 non-null  float64
 4   sqft_living     21613 non-null  int64  
 5   sqft_lot        21613 non-null  int64  
 6   floors          21613 non-null  float64
 7   waterfront      21613 non-null  int64  
 8   condition       21613 non-null  int64  
 9   grade           21613 non-null  int64  
 10  yr_built        21613 non-null  int64  
 11  zipcode         21613 non-null  int64  
 12  price_per_sqft  21613 non-null  float64
dtypes: float64(3), int32(1), int64(9)
memory usage: 2.1 MB


https://www.homedit.com/average-bedroom-size/#:~:text=Average%20Home%20Size%20In%20The%20US&text=Luckily%2C%20there%20is%20a%20fairly,That's%20easy%20math!

Since the average bedroom size is 200 sqft, we want to see the values for those that are less than 200 sqft since that would be abnormally small for a bedroom.

In [26]:
df3[df3.sqft_living/df3.bedrooms<200].head()

id   price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
4203   1437500015  150000         3       0.75          490     38500     1.5   
15870  2402100895  640000        33       1.75         1620      6000     1.0   

       waterfront  condition  grade  yr_built  zipcode  price_per_sqft  
4203            0          4      5      1959    98014      306.122449  
15870           0          5      7      1947    98103      395.061728

https://www.thisoldhouse.com/21017580/8-tiny-houses-that-have-more-storage-than-your-house

I mean again, it's possible but these would be outliers for our analysis.

In [27]:
df3.shape

(21613, 13)

In [28]:
df4 = df3[~(df3.sqft_living/df3.bedrooms<200)]
df4.shape

(21611, 13)

So two of the outliers were removed due to abnormally small bedroom sizes.

Now, let's take a look at price_per_sqft values to observe for outliers, too large or too small.

In [29]:
df4.price_per_sqft.describe()

count    21611.000000
mean       264.148594
std        110.061913
min         87.588235
25%        182.283399
50%        244.609060
75%        318.321796
max        810.138889
Name: price_per_sqft, dtype: float64

## Exploratory Analysis

So far, we have found

## Data Question - Answers:

1. 

## Recommendations:

These findings led us to recommend 